Manifold/Expression.swift

```diff
@@ -333,67 +338,72 @@ extension Expression where Recur: FixpointType, Recur: Equatable {
 			return .right(.Type(n + 1))
 
 		case let .Variable(i):
-			return environment[i].map(Either.Right) ?? Either.Left("Unexpectedly free variable \(i)")
+			return context[i].map(Either.Right) ?? Either.Left("Unexpectedly free variable \(i)")
 
 		case let .Lambda(i, type, body):
-			return type.typecheck(environment, against: .Type(0))
+			return type.typecheck(context, against: .Type(0))
 				.flatMap { _ in
-					body.typecheck(environment + [ .Local(i): type ])
+					body.typecheck(context + [ .Local(i): type ])
 						.map { Expression.lambda(Recur(type), const(Recur($0))) }
-			}
+				}
 
 		case let .Product(a, b):
-			return (a.typecheck(environment) &&& b.typecheck(environment))
+			return (a.typecheck(context) &&& b.typecheck(context))
 				.map { A, B in Expression.lambda(Recur(A), const(Recur(B))) }
 
 		case let .Application(a, b):
-			return a.typecheck(environment)
+			return a.typecheck(context)
 				.flatMap { A in
 					A.analysis(
 						ifLambda: { i, type, body in
-							b.typecheck(environment, against: type.out).map { body.out.substitute(i, $0) }
+							b.typecheck(context, against: type.out).map { body.out.substitute(i, $0) }
 						},
 						otherwise: const(Either.Left("illegal application of \(a) : \(A) to \(b)")))
 				}
 
 		case let .Projection(term, branch):
-			return term.typecheck(environment)
+			return term.typecheck(context)
 				.flatMap { type in
 					type.analysis(
 						ifLambda: { i, A, B in
 							Either.Right(branch ? B.out.substitute(i, A.out) : A.out)
 						},
-						otherwise: const(Either.Left("illegal attempt to project field \(branch ? 1 : 0) of value \(term) of non-product type \(type)")))
-				}
-
-		// Typecheck products annotated with lambda type as sums.
-		case let .Annotation(.Product(tag, body), .Lambda(i, tagType, typeBody)):
-			return tagType.out.typecheck(environment, against: .Type(0))
-				.flatMap { _ in
-					tag.out.typecheck(environment, against: tagType.out)
-						.flatMap { _ in
-							typeBody.out.substitute(i, tag.out).typecheck(environment)
-								.flatMap { bodyType in
-									body.out.typecheck(environment, against: bodyType)
-										.map(const(.Lambda(i, tagType, typeBody)))
-								}
-						}
```

Comment:
This wasn’t quite correct, and, as it turns out, this is the wrong place to employ it anyway.